packages/std/src/errors/std_error.rs

```diff
@@ -0,0 +1,313 @@
+                    && backtraces_eq(backtrace, backtrace2)
+            }
+            _ => false,
+        }
+    }
+}
+
+/// The return type for init, handle and query. Since the error type cannot be serialized to JSON,
+/// this is only available within the contract and its unit tests.
+///
+/// The prefix "Std" means "the standard result within the standard library". This is not the only
+/// result/error type in cosmwasm-std.
+pub type StdResult<T> = core::result::Result<T, StdError>;
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use crate::serde::{from_slice, to_vec};
+
+    #[test]
+    fn can_serialize() {
+        let error = InvalidBase64 {
+            msg: "invalid length".to_string(),
+        }
+        .build();
+        assert_eq!(
+            to_vec(&error).unwrap(),
+            br#"{"invalid_base64":{"msg":"invalid length"}}"#.to_vec()
+        );
+    }
+
+    #[test]
+    fn can_deserialize() {
+        let error: StdError =
+            from_slice(br#"{"invalid_base64":{"msg":"invalid length"}}"#).unwrap();
+        match error {
+            StdError::InvalidBase64 { msg, backtrace } => {
+                assert_eq!(msg, "invalid length");
+                assert!(backtrace.is_none());
+            }
+            _ => panic!("invalid type"),
+        };
+    }
+
+    /// The deseralizer in from_slice can perform zero-copy deserializations (https://serde.rs/lifetimes.html).
+    /// So it is possible to have `&'static str` fields as long as all source data is always static.
+    /// This is an unrealistic assumption for our use case. This test case ensures we can deseralize
+    /// errors from limited liefetime sources.
+    #[test]
+    fn can_deserialize_from_non_static_source() {
+        let source = (br#"{"not_found":{"kind":"bugs"}}"#).to_vec();
+        let error: StdError = from_slice(&source).unwrap();
+        match error {
+            StdError::NotFound { kind, backtrace } => {
+                assert_eq!(kind, "bugs");
+                assert!(backtrace.is_none());
+            }
+            _ => panic!("invalid type"),
+        };
+    }
+
+    #[test]
+    fn eq_works() {
+        let error1 = InvalidBase64 {
+            msg: "invalid length".to_string(),
+        }
+        .build();
+        let error2 = InvalidBase64 {
+            msg: "other bla".to_string(),
+        }
+        .build();
+
+        // Errors are only equal when bactrace is removed
+        let normalized1: StdError = from_slice(&to_vec(&error1).unwrap()).unwrap();
+        let normalized2: StdError = from_slice(&to_vec(&error2).unwrap()).unwrap();
+        assert_ne!(normalized1, normalized2);
+    }
+
+    #[test]
+    fn ne_works() {
+        let error1 = InvalidBase64 {
+            msg: "invalid length".to_string(),
+        }
+        .build();
+        let error2 = InvalidBase64 {
+            msg: "other bla".to_string(),
+        }
+        .build();
+        assert_ne!(error1, error2);
```

Comment:
and this is ne for both backtrace and msg...

Actually, I would ignore the `backtrace` component in PartialEq. And always consider them equal - we are testing the content, not where they were raised - we will never have to identical errors with the same backtrace. To test it out, just update:

```rust
fn backtraces_eq(a: &Option<snafu::Backtrace>, b: &Option<snafu::Backtrace>) -> bool {
     true
}
```

Reply:
I'm trying to develop an idea of what equality means on our errors, independent of how we use it right now. I think I can live with the idea that two errors created at different places of the stack can be equal.

Reply:
Nice